### PR TITLE
Validate block

### LINKED ISSUE #3: Validate mining (block hash and nonce)

### PR DESCRIPTION
**Pull request by cserb**
Fixes #3

- Pow blocks are being validated against nonce and hash
- Pos blocks are being validated against hash

### CODE REVIEW

File: spec/ledger_spec.cr [hunk 1]
```diff
@@ -1,6 +1,33 @@
...
 describe Ledger::Pow do
+  let(:txn) do
+    Ledger::Action::Transaction.new(
+      from: "me",
+      to: "you",
+      amount: 1_u64
+    )
+  end
+  let(:transactions) { [txn] of Ledger::Action::Transaction }
+  let(:block) { Ledger::Pow.mine(transactions) }
```
Comment (q9f) [inline spec/ledger_spec.cr:16]: what does the colon `:` do here? curious about the syntax.

File: src/cocol/node/ledger/api.cr [hunk 2]
```diff
@@ -49,6 +51,11 @@ post "/blocks/pow" do |env|
       response: "Bad Request"
     )
   end
+  halt(
+    env,
+    status_code: 422,
```
Comment (q9f) [inline src/cocol/node/ledger/api.cr:56]: are you defining your own status codes here?

File: src/cocol/node/ledger/api.cr [hunk 3]
```diff
@@ -63,8 +70,22 @@ post "/blocks/pow" do |env|
 end
 
 post "/blocks/pos" do |env|
-  new_block = Ledger::Block::Pos.from_json(
-    env.request.body.not_nil!)
+  begin
+    new_block = Ledger::Block::Pos.from_json(
+      env.request.body.not_nil!)
```
Comment (q9f) [inline src/cocol/node/ledger/api.cr:75]: does `not_nil!` convince the compile this will never fail?

File: spec/ledger_spec.cr (hunk 1 above)
Comment (jkthorne) [reply spec/ledger_spec.cr:16]: it makes the bareword `block` a symbol.
The let function turns it into a method with the name block.

Comment (cserb) [reply spec/ledger_spec.cr:16]: https://github.com/ysbaddaden/minitest.cr

Also it's lazy: http://ruby-journal.com/minitest-let-is-lazy/

File: src/cocol/node/ledger/api.cr (hunk 2 above)
Comment (cserb) [reply src/cocol/node/ledger/api.cr:56]: nope: https://httpstatuses.com/422

File: src/cocol/node/ledger/api.cr (hunk 3 above)
Comment (cserb) [reply src/cocol/node/ledger/api.cr:75]: No it just raises if body is `Nil`
https://crystal-lang.org/api/0.32.1/Object.html#not_nil!-instance-method
https://crystal-lang.org/api/0.32.1/NilAssertionError.html

**Pull request merged**